spanner/spanner_snippets/spanner: TestCustomerManagedEncryptionKeys failed

Note: #2033 was also for this test, but it was closed more than 10 days ago. So, I didn't mark it flaky.

----

commit: 9bc859f48ef701a4b81794b5bb86b509b17e2e4c
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/95c8c738-2c3d-4cd8-ba9a-e84e80c3f971), [Sponge](http://sponge2/95c8c738-2c3d-4cd8-ba9a-e84e80c3f971)
status: failed
<details><summary>Test output</summary><br><pre>/usr/local/go/src/testing/testing.go:1701 +0xe5
/usr/local/go/src/time/sleep.go:180 +0x45
/usr/local/go/src/testing/testing.go:1240 +0x2da
/usr/local/go/src/testing/testing.go:1512 +0x78
/usr/local/go/src/testing/testing.go:1194 +0xef
/usr/local/go/src/testing/testing.go:1510 +0x2fe
/usr/local/go/src/testing/testing.go:1418 +0x1eb
_testmain.go:51 +0x138
/go/pkg/mod/go.opencensus.io@v0.23.0/stats/view/worker.go:276 +0xcd
/go/pkg/mod/go.opencensus.io@v0.23.0/stats/view/worker.go:34 +0x68
/go/pkg/mod/github.com/googleapis/gax-go/v2@v2.0.5/invoke.go:55 +0xa5
/go/pkg/mod/cloud.google.com/go@v0.84.0/longrunning/longrunning.go:156 +0xe5
/go/pkg/mod/cloud.google.com/go@v0.84.0/longrunning/longrunning.go:142 +0xf3
/go/pkg/mod/cloud.google.com/go/spanner@v1.20.0/admin/database/apiv1/database_admin_client.go:1186 +0xaa
/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/spanner_restore_backup_with_encryption_key.go:63 +0x51e
/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/integration_test.go:498 +0x91
/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/integration_test.go:578 +0xb6
/tmpfs/src/github/golang-samples/internal/testutil/retry.go:34 +0xd1
/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/integration_test.go:576 +0x13b
/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/integration_test.go:500 +0xc9b
/usr/local/go/src/testing/testing.go:1194 +0xef
/usr/local/go/src/testing/testing.go:1239 +0x2b3
/go/pkg/mod/google.golang.org/grpc@v1.38.0/balancer_conn_wrappers.go:69 +0xac
/go/pkg/mod/google.golang.org/grpc@v1.38.0/balancer_conn_wrappers.go:60 +0x172
/go/pkg/mod/google.golang.org/grpc@v1.38.0/clientconn.go:1214 +0x465
/go/pkg/mod/google.golang.org/grpc@v1.38.0/clientconn.go:844 +0x12a
/usr/local/go/src/runtime/netpoll.go:222 +0x55
/usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
/usr/local/go/src/internal/poll/fd_unix.go:166 +0x1d5
/usr/local/go/src/net/fd_posix.go:55 +0x4f
/usr/local/go/src/net/net.go:183 +0x91
/usr/local/go/src/crypto/tls/conn.go:776 +0x63
/usr/local/go/src/bytes/buffer.go:204 +0xbe
/usr/local/go/src/crypto/tls/conn.go:798 +0xf3
/usr/local/go/src/crypto/tls/conn.go:605 +0x115
/usr/local/go/src/crypto/tls/conn.go:573
/usr/local/go/src/crypto/tls/conn.go:1276 +0x165
/usr/local/go/src/bufio/bufio.go:227 +0x222
/usr/local/go/src/io/io.go:328 +0x87
/usr/local/go/src/io/io.go:347
/go/pkg/mod/golang.org/x/net@v0.0.0-20210503060351-7fd8e65b6420/http2/frame.go:237 +0x89
/go/pkg/mod/golang.org/x/net@v0.0.0-20210503060351-7fd8e65b6420/http2/frame.go:492 +0xa5
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:1347 +0x1a5
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:346 +0xdd1
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/controlbuf.go:407 +0xff
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/controlbuf.go:527 +0x1dd
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:396 +0x7b
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:394 +0x12ae
/usr/local/go/src/runtime/netpoll.go:222 +0x55
/usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
/usr/local/go/src/internal/poll/fd_unix.go:166 +0x1d5
/usr/local/go/src/net/fd_posix.go:55 +0x4f
/usr/local/go/src/net/net.go:183 +0x91
/usr/local/go/src/crypto/tls/conn.go:776 +0x63
/usr/local/go/src/bytes/buffer.go:204 +0xbe
/usr/local/go/src/crypto/tls/conn.go:798 +0xf3
/usr/local/go/src/crypto/tls/conn.go:605 +0x115
/usr/local/go/src/crypto/tls/conn.go:573
/usr/local/go/src/crypto/tls/conn.go:1276 +0x165
/usr/local/go/src/bufio/bufio.go:227 +0x222
/usr/local/go/src/io/io.go:328 +0x87
/usr/local/go/src/io/io.go:347
/go/pkg/mod/golang.org/x/net@v0.0.0-20210503060351-7fd8e65b6420/http2/frame.go:237 +0x89
/go/pkg/mod/golang.org/x/net@v0.0.0-20210503060351-7fd8e65b6420/http2/frame.go:492 +0xa5
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:1347 +0x1a5
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:346 +0xdd1
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/controlbuf.go:407 +0xff
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/controlbuf.go:527 +0x1dd
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:396 +0x7b
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:394 +0x12ae
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/controlbuf.go:407 +0xff
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/controlbuf.go:527 +0x1dd
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:396 +0x7b
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:394 +0x12ae
/go/pkg/mod/google.golang.org/grpc@v1.38.0/balancer_conn_wrappers.go:69 +0xac
/go/pkg/mod/google.golang.org/grpc@v1.38.0/balancer_conn_wrappers.go:60 +0x172
/go/pkg/mod/google.golang.org/grpc@v1.38.0/clientconn.go:1214 +0x465
/go/pkg/mod/google.golang.org/grpc@v1.38.0/clientconn.go:844 +0x12a
/usr/local/go/src/runtime/netpoll.go:222 +0x55
/usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
/usr/local/go/src/internal/poll/fd_unix.go:166 +0x1d5
/usr/local/go/src/net/fd_posix.go:55 +0x4f
/usr/local/go/src/net/net.go:183 +0x91
/usr/local/go/src/crypto/tls/conn.go:776 +0x63
/usr/local/go/src/bytes/buffer.go:204 +0xbe
/usr/local/go/src/crypto/tls/conn.go:798 +0xf3
/usr/local/go/src/crypto/tls/conn.go:605 +0x115
/usr/local/go/src/crypto/tls/conn.go:573
/usr/local/go/src/crypto/tls/conn.go:1276 +0x165
/usr/local/go/src/bufio/bufio.go:227 +0x222
/usr/local/go/src/io/io.go:328 +0x87
/usr/local/go/src/io/io.go:347
/go/pkg/mod/golang.org/x/net@v0.0.0-20210503060351-7fd8e65b6420/http2/frame.go:237 +0x89
/go/pkg/mod/golang.org/x/net@v0.0.0-20210503060351-7fd8e65b6420/http2/frame.go:492 +0xa5
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:1347 +0x1a5
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:346 +0xdd1
/go/pkg/mod/google.golang.org/grpc@v1.38.0/clientconn.go:1214 +0x465
/go/pkg/mod/google.golang.org/grpc@v1.38.0/clientconn.go:844 +0x12a
/go/pkg/mod/google.golang.org/grpc@v1.38.0/balancer_conn_wrappers.go:69 +0xac
/go/pkg/mod/google.golang.org/grpc@v1.38.0/balancer_conn_wrappers.go:60 +0x172
/go/pkg/mod/google.golang.org/grpc@v1.38.0/clientconn.go:1214 +0x465
/go/pkg/mod/google.golang.org/grpc@v1.38.0/clientconn.go:844 +0x12a
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/controlbuf.go:407 +0xff
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/controlbuf.go:527 +0x1dd
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:396 +0x7b
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:394 +0x12ae
/usr/local/go/src/runtime/netpoll.go:222 +0x55
/usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
/usr/local/go/src/internal/poll/fd_unix.go:166 +0x1d5
/usr/local/go/src/net/fd_posix.go:55 +0x4f
/usr/local/go/src/net/net.go:183 +0x91
/usr/local/go/src/crypto/tls/conn.go:776 +0x63
/usr/local/go/src/bytes/buffer.go:204 +0xbe
/usr/local/go/src/crypto/tls/conn.go:798 +0xf3
/usr/local/go/src/crypto/tls/conn.go:605 +0x115
/usr/local/go/src/crypto/tls/conn.go:573
/usr/local/go/src/crypto/tls/conn.go:1276 +0x165
/usr/local/go/src/bufio/bufio.go:227 +0x222
/usr/local/go/src/io/io.go:328 +0x87
/usr/local/go/src/io/io.go:347
/go/pkg/mod/golang.org/x/net@v0.0.0-20210503060351-7fd8e65b6420/http2/frame.go:237 +0x89
/go/pkg/mod/golang.org/x/net@v0.0.0-20210503060351-7fd8e65b6420/http2/frame.go:492 +0xa5
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:1347 +0x1a5
/go/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_client.go:346 +0xdd1
/go/pkg/mod/google.golang.org/grpc@v1.38.0/balancer_conn_wrappers.go:69 +0xac
/go/pkg/mod/google.golang.org/grpc@v1.38.0/balancer_conn_wrappers.go:60 +0x172</pre></details>

It seems that the entire test run timed out while executing a restore operation for Spanner:

```
panic: test timed out after 1h0m0s

goroutine 7945 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1701 +0xe5
created by time.goFunc
	/usr/local/go/src/time/sleep.go:180 +0x45

goroutine 1 [chan receive, 58 minutes]:
testing.(*T).Run(0xc00086b680, 0xc4d114, 0x21, 0xc6d0c8, 0x498101)
	/usr/local/go/src/testing/testing.go:1240 +0x2da
testing.runTests.func1(0xc0003f8000)
	/usr/local/go/src/testing/testing.go:1512 +0x78
testing.tRunner(0xc0003f8000, 0xc00035fde0)
	/usr/local/go/src/testing/testing.go:1194 +0xef
testing.runTests(0xc00000ee10, 0x115dd20, 0x5, 0x5, 0xc02bab59c2792d5f, 0x3463109bb05, 0x116d360, 0xc3eb60)
	/usr/local/go/src/testing/testing.go:1510 +0x2fe
testing.(*M).Run(0xc000196a00, 0x0)
	/usr/local/go/src/testing/testing.go:1418 +0x1eb
main.main()
	_testmain.go:51 +0x138

goroutine 6 [select]:
go.opencensus.io/stats/view.(*worker).start(0xc000196580)
	/go/pkg/mod/go.opencensus.io@v0.23.0/stats/view/worker.go:276 +0xcd
created by go.opencensus.io/stats/view.init.0
	/go/pkg/mod/go.opencensus.io@v0.23.0/stats/view/worker.go:34 +0x68

goroutine 7474 [select, 2 minutes]:
github.com/googleapis/gax-go/v2.Sleep(0xd23848, 0xc0005ff1a0, 0x99cbd10d1, 0xd21890, 0xc00075c5a0)
	/go/pkg/mod/github.com/googleapis/gax-go/v2@v2.0.5/invoke.go:55 +0xa5
cloud.google.com/go/longrunning.(*Operation).wait(0xc000258078, 0xd23848, 0xc0005ff1a0, 0xd21890, 0xc00075c5a0, 0xc000525950, 0xc6d368, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/cloud.google.com/go@v0.84.0/longrunning/longrunning.go:156 +0xe5
cloud.google.com/go/longrunning.(*Operation).WaitWithInterval(0xc000258078, 0xd23848, 0xc0005ff1a0, 0xd21890, 0xc00075c5a0, 0xdf8475800, 0x0, 0x0, 0x0, 0xb04700, ...)
	/go/pkg/mod/cloud.google.com/go@v0.84.0/longrunning/longrunning.go:142 +0xf3
cloud.google.com/go/spanner/admin/database/apiv1.(*RestoreDatabaseOperation).Wait(0xc00045e020, 0xd23848, 0xc0005ff1a0, 0x0, 0x0, 0x0, 0x0, 0xc00045e020, 0x0)
	/go/pkg/mod/cloud.google.com/go/spanner@v1.20.0/admin/database/apiv1/database_admin_client.go:1186 +0xaa
github.com/GoogleCloudPlatform/golang-samples/spanner/spanner_snippets/spanner.restoreBackupWithCustomerManagedEncryptionKey(0xd23848, 0xc0005ff1a0, 0xd13c60, 0xc000434f00, 0xc000156070, 0x63, 0xc000256220, 0x1e, 0xc0001520e0, 0x6f, ...)
	/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/spanner_restore_backup_with_encryption_key.go:63 +0x51e
github.com/GoogleCloudPlatform/golang-samples/spanner/spanner_snippets/spanner.TestCustomerManagedEncryptionKeys.func3(0xd23848, 0xc0005ff1a0, 0xd13c60, 0xc000434f00, 0xc000156070, 0x63, 0xc000256220, 0x1e, 0x7fe81883bf58, 0x8)
	/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/integration_test.go:498 +0x91
github.com/GoogleCloudPlatform/golang-samples/spanner/spanner_snippets/spanner.runBackupSampleWithRetry.func1(0xc00012c000)
	/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/integration_test.go:578 +0xb6
github.com/GoogleCloudPlatform/golang-samples/internal/testutil.Retry(0xc00086b680, 0xa, 0xdf8475800, 0xc000525cf8, 0x11559a0)
	/tmpfs/src/github/golang-samples/internal/testutil/retry.go:34 +0xd1
github.com/GoogleCloudPlatform/golang-samples/spanner/spanner_snippets/spanner.runBackupSampleWithRetry(0xd23848, 0xc0005ff1a0, 0xc00086b680, 0xc0001dbea8, 0xc000156070, 0x63, 0xc000256220, 0x1e, 0xc5e797, 0x3f, ...)
	/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/integration_test.go:576 +0x13b
github.com/GoogleCloudPlatform/golang-samples/spanner/spanner_snippets/spanner.TestCustomerManagedEncryptionKeys(0xc00086b680)
	/tmpfs/src/github/golang-samples/spanner/spanner_snippets/spanner/integration_test.go:500 +0xc9b
testing.tRunner(0xc00086b680, 0xc6d0c8)
	/usr/local/go/src/testing/testing.go:1194 +0xef
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1239 +0x2b3
```


There are however multiple other test failures, all related to setting / changing permissions, but none are for Spanner:

```
=== RUN   TestDatasets
    integration_test.go:54: updateDataSetAccessControl("golang_snippettest_dataset_07379da726c745d080917942d385eee5"): googleapi: Error 403: IAM setPolicy failed for Dataset golang-samples-tests-6:golang_snippettest_dataset_07379da726c745d080917942d385eee5: One or more users named in the policy do not belong to a permitted customer., policyViolation
--- FAIL: TestDatasets (6.49s)
FAIL
FAIL	github.com/GoogleCloudPlatform/golang-samples/bigquery/snippets/dataset	6.511s
```

```
=== RUN   TestPolicyTagManager
    integration_test.go:87: setIAMPolicy(projects/golang-samples-tests-6/locations/us/taxonomies/1520077384788401495/policyTags/8517990187981738822): SetIamPolicy: rpc error: code = FailedPrecondition desc = One or more users named in the policy do not belong to a permitted customer.
    integration_test.go:95: unexpected output ("Testing the permissions on projects/golang-samples-tests-6/locations/us/taxonomies/1520077384788401495/policyTags/8517990187981738822, of the 1 permissions probed, caller has 0 permissions\n") did not contain (of the 1 permissions probed, caller has 1 permissions: datacatalog.categories.fineGrainedGet)
--- FAIL: TestPolicyTagManager (1.19s)
FAIL
FAIL	github.com/GoogleCloudPlatform/golang-samples/datacatalog/snippets/policytagmanager	1.194s
```

PubSub:
```
=== RUN   TestIAM
    retry.go:44: FAILED after 10 attempts:
        subscription_test.go:184: addUsers: SetPolicy: rpc error: code = FailedPrecondition desc = One or more users named in the policy do not belong to a permitted customer.
    retry.go:44: FAILED after 10 attempts:
        subscription_test.go:195: want "group:cloud-logs@google.com" as viewer, policy=&{etag:"\x00 \x01"}
        subscription_test.go:198: want "allUsers" as viewer, policy=&{etag:"\x00 \x01"}
--- FAIL: TestIAM (24.23s)
```

```
=== RUN   TestIAM
    retry.go:44: FAILED after 10 attempts:
        topics_test.go:196: addUsers: SetPolicy: rpc error: code = FailedPrecondition desc = One or more users named in the policy do not belong to a permitted customer.
    retry.go:44: FAILED after 10 attempts:
        topics_test.go:207: want "group:cloud-logs@google.com" as viewer, policy=&{etag:"\x00 \x01"}
        topics_test.go:210: want "allUsers" as viewer, policy=&{etag:"\x00 \x01"}
--- FAIL: TestIAM (23.60s)
```

```
=== RUN   TestACL
    acl_test.go:59: addBucketOwner: ACLHandle.Set: googleapi: Error 412: One or more users named in the policy do not belong to a permitted customer., conditionNotMet
    acl_test.go:74: removeBucketOwner: ACLHandle.Delete: googleapi: Error 404: The specified key does not exist., notFound
--- FAIL: TestACL (3.48s)
FAIL
FAIL	github.com/GoogleCloudPlatform/golang-samples/storage/acl	3.482s
```

```
=== RUN   TestSetBucketPublicIAM
    buckets_test.go:522: setBucketPublicIAM: Bucket("golang-samples-tests-6-storage-buckets-tests").IAM().SetPolicy: googleapi: Error 412: One or more users named in the policy do not belong to a permitted customer., conditionNotMet
--- FAIL: TestSetBucketPublicIAM (3.09s)
=== RUN   TestDelete
--- PASS: TestDelete (2.13s)
FAIL
FAIL	github.com/GoogleCloudPlatform/golang-samples/storage/buckets	35.653s
```

```
=== CONT  TestHTTP2Server
    h2c.e2e_test.go:59: http2.Get: unexpected response status: 403 Forbidden
2021/06/19 18:31:42 Attempt #1: Running: operation [delete service] for service [h2c]...
2021/06/19 18:31:42 Attempt #1: Executing: operation [delete service] for service [h2c]: /tmp/google-cloud-sdk/bin/gcloud: --quiet run services delete h2c-20210619-182730 --project golang-samples-tests-6 --platform managed --region us-central1
2021/06/19 18:31:48 Attempt #1: Running: operation [delete container image] for service [h2c]...
2021/06/19 18:31:48 Attempt #1: Executing: operation [delete container image] for service [h2c]: /tmp/google-cloud-sdk/bin/gcloud: --quiet container images delete gcr.io/golang-samples-tests-6/h2c:20210619-182730
--- FAIL: TestHTTP2Server (76.75s)
```

@olavloite I think I have found the reason. For the test, the DB names are different for the backup and the restore. 

I ran the above two tests locally and get the following error (with some debugging printouts): 

```
=== RUN   TestBackupSample
=== PAUSE TestBackupSample
=== RUN   TestBackupCustomerManagedEncryptionKeys
=== PAUSE TestBackupCustomerManagedEncryptionKeys
=== CONT  TestBackupSample
=== CONT  TestBackupCustomerManagedEncryptionKeys
TestBackupSample dbName:projects/[hide-project-id]/instances/go-hengfeng-97b30181-a0e5-46/databases/smpl-1626055810-f168a628
TestCustomerManagedEncryptionKeys dbName:projects/[hide-project-id]/instances/go-hengfeng-dcac226b-f568-49/databases/smpl-1626055810-431d8206
    integration_test.go:522: failed to restore database with customer managed encryption key: restoreBackupWithCustomerManagedEncryptionKey.RestoreDatabase: rpc error: code = NotFound desc = Backup not found: projects/[hide-project-id]/instances/go-hengfeng-465c6a42-1d7c-41/backups/enc-backup-1626055842-4488037d
--- FAIL: TestBackupCustomerManagedEncryptionKeys (275.10s)
=== CONT  TestBackupSample
    integration_test.go:522: failed to restore a backup: rpc error: code = NotFound desc = Backup not found: projects[hide-project-id]/instances/go-hengfeng-5f1b9581-8df3-44/backups/backup-1626055810-f168a628
--- FAIL: TestBackupSample (304.73s)
FAIL
FAIL    github.com/GoogleCloudPlatform/golang-samples/spanner/spanner_snippets/spanner  304.944s
FAIL
```

As you can see, all of their instance IDs are different. 

https://github.com/GoogleCloudPlatform/golang-samples/blob/03b672e88838a8d50647cd111c05d81ef20ff488/spanner/spanner_snippets/spanner/integration_test.go#L325-L327

`dbName` and `restoreDBName` are not in the same instance. When we restore the DB, we use `restoreDBName` and `backupID` so that we can't find the backup.

--- a/spanner/spanner_snippets/spanner/integration_test.go
+++ b/spanner/spanner_snippets/spanner/integration_test.go
@@ -50,11 +50,11 @@ var (
 	validInstancePattern = regexp.MustCompile("^projects/(?P<project>[^/]+)/instances/(?P<instance>[^/]+)$")
 )
 
-func initTest(t *testing.T, id string) (dbName string, cleanup func()) {
+func initTest(t *testing.T, id string) (instName, dbName string, cleanup func()) {
 	projectID := getSampleProjectId(t)
-	instance, cleanup := createTestInstance(t, projectID)
+	instName, cleanup = createTestInstance(t, projectID)
 	dbID := validLength(fmt.Sprintf("smpl-%s", id), t)
-	dbName = fmt.Sprintf("%s/databases/%s", instance, dbID)
+	dbName = fmt.Sprintf("%s/databases/%s", instName, dbID)
 
 	return
 }
@@ -83,11 +83,9 @@ func getVersionTime(t *testing.T, dbName string) (versionTime time.Time) {
 	return versionTime
 }
 
-func initBackupTest(t *testing.T, id string) (restoreDBName, backupID, cancelledBackupID string, cleanup func()) {
-	projectID := getSampleProjectId(t)
-	instance, cleanup := createTestInstance(t, projectID)
+func initBackupTest(t *testing.T, id, instName string) (restoreDBName, backupID, cancelledBackupID string) {
 	restoreDatabaseID := validLength(fmt.Sprintf("restore-%s", id), t)
-	restoreDBName = fmt.Sprintf("%s/databases/%s", instance, restoreDatabaseID)
+	restoreDBName = fmt.Sprintf("%s/databases/%s", instName, restoreDatabaseID)
 	backupID = validLength(fmt.Sprintf("backup-%s", id), t)
 	cancelledBackupID = validLength(fmt.Sprintf("cancel-%s", id), t)
 
@@ -116,7 +114,7 @@ func TestSample(t *testing.T) {
 	_ = testutil.SystemTest(t)
 	t.Parallel()
 
-	dbName, cleanup := initTest(t, randomID())
+	_, dbName, cleanup := initTest(t, randomID())
 	defer cleanup()
 
 	ctx, cancel := context.WithTimeout(context.Background(), time.Hour)
@@ -322,9 +320,9 @@ func TestBackupSample(t *testing.T) {
 	t.Parallel()
 
 	id := randomID()
-	dbName, cleanup := initTest(t, id)
+	instName, dbName, cleanup := initTest(t, id)
 	defer cleanup()
-	restoreDBName, backupID, cancelledBackupID, cleanupBackup := initBackupTest(t, id)
+	restoreDBName, backupID, cancelledBackupID := initBackupTest(t, id, instName)
 
 	var out string
 	// Set up the database for testing backup operations.
@@ -358,9 +356,6 @@ func TestBackupSample(t *testing.T) {
 	out = runSampleWithContext(ctx, t, listDatabaseOperations, restoreDBName, "failed to list database operations")
 	assertContains(t, out, fmt.Sprintf("Database %s restored from backup", restoreDBName))
 
-	// Delete the restore DB.
-	cleanupBackup()
-
 	out = runBackupSample(ctx, t, deleteBackup, dbName, backupID, "failed to delete a backup")
 	assertContains(t, out, fmt.Sprintf("Deleted backup %s", backupID))
 }
@@ -369,7 +364,7 @@ func TestCreateDatabaseWithRetentionPeriodSample(t *testing.T) {
 	_ = testutil.SystemTest(t)
 	t.Parallel()
 
-	dbName, cleanup := initTest(t, randomID())
+	_, dbName, cleanup := initTest(t, randomID())
 	defer cleanup()
 
 	ctx, cancel := context.WithTimeout(context.Background(), time.Hour)
@@ -386,14 +381,11 @@ func TestCustomerManagedEncryptionKeys(t *testing.T) {
 	tc := testutil.SystemTest(t)
 	t.Parallel()
 
-	dbName, cleanup := initTest(t, randomID())
+	instName, dbName, cleanup := initTest(t, randomID())
 	defer cleanup()
 
 	var b bytes.Buffer
 
-	projectID := getSampleProjectId(t)
-	instanceName, cleanup := createTestInstance(t, projectID)
-	defer cleanup()
 	locationId := "us-central1"
 	keyRingId := "spanner-test-keyring"
 	keyId := "spanner-test-key"
@@ -431,7 +423,7 @@ func TestCustomerManagedEncryptionKeys(t *testing.T) {
 	assertContains(t, out, fmt.Sprintf("using encryption key %s", kmsKeyName))
 
 	// Try to restore the encrypted database and delete the restored database after the test.
-	restoredName := fmt.Sprintf("%s/databases/rest-enc-%s", instanceName, randomID())
+	restoredName := fmt.Sprintf("%s/databases/rest-enc-%s", instName, randomID())
 	restoreFunc := func(ctx context.Context, w io.Writer, dbName, backupID string) error {
 		return restoreBackupWithCustomerManagedEncryptionKey(ctx, w, dbName, backupId, kmsKeyName)
 	}

